### ISSUE TITLE
SPLA hangs on a 6x6 grid

### ISSUE THREAD

**Issue by toxa81** [eth-cscs/spla#22]
For this test case https://github.com/electronic-structure/SIRIUS/blob/develop/apps/tests/test_wf_inner.cpp spla will hang with the following submission

```
#!/bin/bash -l
#SBATCH --job-name="test_scf"
#SBATCH --nodes=36
#SBATCH --time=00:04:00
#SBATCH --account=csstaff
#SBATCH -C gpu

export MPICH_MAX_THREAD_SAFETY=multiple
export CRAY_CUDA_MPS=0
export MKL_NUM_THREADS=1
export OMP_NUM_THREADS=1

srun -N 36 -n 36 --hint=nomultithread --unbuffered -c 1 ./test_wf_inner  --M=2248 --N=2248 --BS=128 --mpi_grid=6:6
```
Number of OMP threads does not play a role. With 12 threads it also hangs.

**Comment by AdhocMan** [comment]
Thanks for providing the example. I managed to reproduce the problem and I think I've identified the issue.
Internally, SPLA tries to find an optimal block size for processing, which depends on the input size, input location and number of MPI ranks. When a nulltptr is passed as input only on some ranks, while device memory is passed in on others, the internal block sizes can differ. Since the number of MPI ranks and input size also plays a role, this only happens in specific configurations.
I've pushed a fix and I'll create a new release shortly.
